robot-server/robot_server/protocols/protocol_store.py

```
@@ -26,44 +33,137 @@ def __init__(self, protocol_id: str) -> None:
-        """Remove a protocol from the store."""
-        try:
-            entry = self._protocols_by_id.pop(protocol_id)
-        except KeyError as e:
-            raise ProtocolNotFoundError(protocol_id) from e
-
-        try:
-            protocol_dir = entry.source.directory
-            for file_path in entry.source.files:
-                (protocol_dir / file_path.name).unlink()
-            protocol_dir.rmdir()
-        except Exception as e:
-            log.warning(
-                f"Unable to delete all files for protocol {protocol_id}",
-                exc_info=e,
-            )
-
-        return entry
+        """Remove a `ProtocolResource` from the store.
+
+        After removing it from the store, attempt to delete all files that it
+        referred to.
+
+        Returns:
+            The `ProtocolResource` as it appeared just before removing it.
+            Note that the files it refers to will no longer exist.
+        """
+        select_statement = sqlalchemy.select(_protocol_table).where(
+            _protocol_table.c.id == protocol_id
+        )
+        delete_statement = sqlalchemy.delete(_protocol_table).where(
+            _protocol_table.c.id == protocol_id
+        )
+        with self._sql_engine.begin() as transaction:
+            try:
+                # SQLite <3.35.0 doesn't support the RETURNING clause,
+                # so we do it ourselves with a separate SELECT.
+                row_to_delete = transaction.execute(select_statement).one()
+            except sqlalchemy.exc.NoResultFound as e:
+                raise ProtocolNotFoundError(protocol_id=protocol_id) from e
+            transaction.execute(delete_statement)
+
+        deleted_resource = _convert_sql_row_to_resource(sql_row=row_to_delete)
+
+        protocol_dir = deleted_resource.source.directory
+        for file_path in deleted_resource.source.files:
+            (protocol_dir / file_path.name).unlink()
+        protocol_dir.rmdir()
```

I removed the exception swallowing from this block that attempts to delete all files. If something here doesn't work, that's a legitimate server-side error. We want integration tests to catch it, in particular.